### PR TITLE
Mass COM calculation

### PR DESCRIPTION
**Pull request by richardjgowers**
So over in #670 we mooted having a mass COM method, so I mocked this up quickly in Cython. (it's roughly 30x faster than previous)

Basically, given a list of 0 based resids, returns the center of geometry for each resid. So is the equivalent of:
`np.array([r.center_of_geometry() for r in u.residues])`

So there's lots of rough edges as to how the function works, including adding a mass (and charge?) weighting option so it returns CoM.

So my question is, is this as good as cython as we can write? (maybe @kain88-de or anyone better at cython than me can say)

I did a numpy typed version:
https://github.com/MDAnalysis/mdanalysis/commit/12060928efe53d6f48454d8ab8a197488664682c

And a memory view version:
https://github.com/MDAnalysis/mdanalysis/commit/62fdf9927f3a9187df5670703b5a027f9c726338

And couldn't see any difference in the performance

### CODE REVIEW

File: package/MDAnalysis/lib/trans.pyx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+cimport cython
+cimport numpy as np
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def residue_positions(
+        float[:, :] coords not None,
```
Comment (kain88-de) [inline package/MDAnalysis/lib/trans.pyx]: You can tell the compiler how the array is stored (Row or Column Major). That can be done by giving it the strides directly `float[:, ::1` is row major for example the C/Python standard. That should give a small boost but requires that the given arrays follow that memory layout.

File: package/MDAnalysis/lib/trans.pyx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        long[:] indices not None):
+    cdef Py_ssize_t i, j, k
```
Comment (kain88-de) [inline package/MDAnalysis/lib/trans.pyx]: I usually use just int here as type.

File: package/MDAnalysis/lib/trans.pyx [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+cimport cython
+cimport numpy as np
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def residue_positions(
```
Comment (kain88-de) [inline package/MDAnalysis/lib/trans.pyx]: a doc string would be nice.

**Comment by kain88-de** [comment]
What was your benchmark? I don't see anything greatly wrong here so it should be fine and generally faster then a python version. 

Fix the doc string and it should be good to go.

File: package/MDAnalysis/lib/trans.pyx (hunk 1 above)
Comment (kain88-de) [inline package/MDAnalysis/lib/trans.pyx]: I would keep your current version as this is less pain for someone who uses the function.

File: package/MDAnalysis/lib/trans.pyx [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+cimport cython
+cimport numpy as np
+
+
+@cython.boundscheck(False)
```
Comment (hainm) [inline package/MDAnalysis/lib/trans.pyx]: you can turn this off globally in setup.py file. check my pytraj's one

https://github.com/Amber-MD/pytraj/blob/master/setup.py#L327-L347

**Comment by richardjgowers** [comment]
@kain88-de I'm loading the PSF/DCD system and doing 
``` python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe('adk.psf','adk_dims.dcd')

In [3]: from MDAnalysis.lib.trans import residue_positions

In [4]: indices = u.atoms.resids - 1

In [5]: %timeit out = residue_positions(u.atoms.positions, indices)
10000 loops, best of 3: 140 µs per loop

In [6]: %timeit long_way = np.array([r.center_of_geometry() for r in u.residues])
100 loops, best of 3: 3.03 ms per loop
```

I'll look at defining the layout and see..

**Comment by kain88-de** [comment]
You don't need to define the layout. That was more as an FYI. It should be absolutely fine here as it is.

**Pull request closed without merging**